src/StructuredYson/StructuredYson.tsx

```diff
@@ -259,36 +274,76 @@ export class StructuredYson extends React.PureComponent<Props, State> {
-                this.searchRef.current?.focus();
+            nextSlash++;
+        }
+        // Check the full path as well
+        if (collapsedState[targetMatchPath]) {
+            if (!hasCollapsedParents) {
+                effectiveCollapsedState = {...collapsedState};
+                hasCollapsedParents = true;
             }
+            delete effectiveCollapsedState[targetMatchPath];
+        }
+
+        // If we expanded any parents, recalculate state and retry
+        if (hasCollapsedParents) {
+            this.updateState({collapsedState: effectiveCollapsedState}, () => {
+                // After state recalculation, find the target in the new allMatchPaths
+                const newAllMatchPaths = this.state.allMatchPaths;
+                const newTargetIndex = newAllMatchPaths.indexOf(targetMatchPath);
+                if (newTargetIndex !== -1) {
+                    // Navigate to the target using its new index
+                    const newMatchedRows = this.state.matchedRows;
+                    let visibleMatchCount = 0;
+                    for (let i = 0; i < newTargetIndex; i++) {
+                        if (
+                            this.findCollapsedParent(
+                                newAllMatchPaths[i],
+                                this.state.collapsedState,
+                            ) === null
+                        ) {
+                            visibleMatchCount++;
+                        }
+                    }
+                    if (visibleMatchCount < newMatchedRows.length) {
+                        this.setState({matchIndex: newTargetIndex});
+                        this.tableRef.current?.scrollToIndex(newMatchedRows[visibleMatchCount]);
+                        this.searchRef.current?.focus();
+                    }
+                }
+            });
             return;
         }
 
-        // Target is hidden - expand the collapsed parent
-        const newCollapsedState = {...collapsedState};
-        delete newCollapsedState[collapsedParent];
+        // Target is visible - navigate to it
+        // Count how many visible matches are before our target
+        let visibleMatchCount = 0;
+        for (let i = 0; i < nextTotalIndex; i++) {
+            if (this.findCollapsedParent(allMatchPaths[i], effectiveCollapsedState) === null) {
+                visibleMatchCount++;
+            }
+        }
 
-        // Recalculate state with new collapsed state
-        this.updateState({collapsedState: newCollapsedState, matchIndex: nextTotalIndex}, () => {
-            // Retry navigation to the same target
-            this.onNextMatch(null, 0);
-        });
+        // Navigate to the visible match
+        if (visibleMatchCount < matchedRows.length) {
+            this.setState({matchIndex: nextTotalIndex});
+            this.tableRef.current?.scrollToIndex(matchedRows[visibleMatchCount]);
+            this.searchRef.current?.focus();
+        }
```

Comment:
Line ranges 310-326 and 334-346 looks like a copy paste. Isn't it better to extract it to a local helper, like:
```ts
const navigateTo = ({matchIndex, matchedRows, collapsedState}) => {
    ...
}
```
Is there any blockers for such approach?